scipy/optimize/_zeros_py.py

```diff
@@ -2866,3 +2866,240 @@ def _differentiate_weights(work, n):
+            - ``-2`` : The maximum number of iterations was reached.
+            - ``-3`` : A non-finite value was encountered.
+            - ``-4`` : Iteration was terminated by `callback`.
+            - ``1`` : The algorithm is proceeding normally (in `callback` only).
+
+        success : bool
+            ``True`` when the algorithm terminated successfully (status ``0``).
+
+    Notes
+    -----
+    Similar to `scipy.optimize.bracket`, this function seeks to find real
+    points xl < xm < xr such that f(xl) >= f(xm) and f(xr) >= f(xm), where at
+    least one of the inequalities is strict. Unlike `scipy.optimize.bracket`,
+    this function can operate in a vectorized manner on array input, so long as
+    the input arrays are broadcastable with each other. Also unlike
+    `scipy.optimize.bracket`, users may specify minimum and maximum endpoints
+    for the desired bracket.
+
+    Given an initial trio of points `xl`, `xm`, `xr`, the algorithm checks if
+    these points already give a valid bracket. If not, a new endpoint, `w` is
+    chosen in the "downhill" direction, `xm` becomes the new opposite endpoint,
+    and either `xl` or `xr` becomes the new middle point, depending on which
+    direction is downhill. The algorithm repeats from here.
+
+    The new endpoint `w` is chosen differently depending on whether or not a
+    boundary `min` or `max` has been set in the downhill direction. Without
+    loss of generality, suppose the downhill direction is to the right, so that
+    ``f(xl) > f(xm) > f(xr)``. If there is no boundary to the right, then `w`
+    is chosen to be ``xr + phi * (xr - xm)`` where `phi` is the golden ratio;
+    so that step sizes increase in geometric proportion, with each step size
+    equal to the sum of the two previous step sizes. If there is a boundary,
+    `max` in this case, then `w` is chosen to be ``xr + (max - xr)/phi``, so
+    that step sizes decrease in geometric proportion, slowing to a stop at
+    `max`. This cautious approach ensures that a minima near but distinct from
+    the boundary isn't missed.
+    """  #noqa: E501
+    callback = None  # works; I just don't want to test it
+    temp = _bracket_minima_iv(func, a, b, c, min, max, args, maxiter)
+    func, a, b, c, min, max, args, maxiter = temp
+
+    xs = (a, b, c)
+    xs, fs, args, shape, dtype = _scalar_optimization_initialize(func, xs, args)
+
+    a, b, c = xs
+    fa, fb, fc = fs
+    min = np.broadcast_to(min, shape).astype(dtype, copy=False).ravel()
+    max = np.broadcast_to(max, shape).astype(dtype, copy=False).ravel()
+
+    # To simplify the logic, swap a and b if f(a) < f(b). We should always be marching
+    # downhill in the direction from b to a.
+    comp = fa < fb
+    a[comp], b[comp] = b[comp], a[comp]
+    fa[comp], fb[comp] = fb[comp], fa[comp]
+    # We only need the boundary in the direction we're traveling.
+    limit = np.where(comp, min, max)
+
+    golden_ratio = 1.618033988749895
```

Comment:
Despite this, the code seems to respect `dtype`, probably because `step` has the `dtype` of `a`. But it might be better to explicitly cast to `dtype` to avoid conversion each iteration. I don't think there are any other constants to worry about.